### PR TITLE
Adds configuration flag for the deprecation of associate_user

### PR DESCRIPTION
**Pull request by carlosep**
**Description**
This Pull Request solves the issue https://github.com/solidusio/solidus/issues/3559, and it is done after https://github.com/solidusio/solidus/pull/3667, which was closed due to a better solution on this one.

Based on [this discussion](https://github.com/solidusio/solidus/pull/3494#discussion_r386117331) the `associate_user` method is deprecated due to changes during the creation or building of an order via `current_order` where before the creation and association of an user were made in different places now it is done immediately. After discussing on the issue we came to the conclusion that a good solution would be to add a configuration flag and enter a deprecation cycle starting at 2.11. For that I added the flag `:associate_user_in_authentication_extension` which will allow clients to safely move their customization on `associate_user`

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [x] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: core/lib/spree/core/controller_helpers/order.rb [hunk 1]
```diff
@@ -52,6 +52,8 @@ def current_order(options = {})
         end
 
         def associate_user
+          return if Spree::Config.manually_associate_user
+          warn "Order#associate_user is deprecated and will be removed in 2.11"
```
Comment (elia) [inline core/lib/spree/core/controller_helpers/order.rb:56]: Great! All deprecation in Solidus are managed via `Spree::Deprecation.warn` (see examples around the code) and would also be better to have a spec ensuring the deprecation is displayed (there are numerous examples of that in spec/ as well).

File: core/lib/spree/app_configuration.rb [hunk 2]
```diff
@@ -173,6 +173,10 @@ class AppConfiguration < Preferences::Configuration
     #   @return [String] Email address used as +From:+ field in transactional emails.
     preference :mails_from, :string, default: 'spree@example.com'
 
+    # @!attribute [rw] manually_associate_user
+    #   @return [Boolean] manually call associate_user instead of relying on current_order
+    preference :manually_associate_user, :boolean, default: true
```
Comment (elia) [inline core/lib/spree/app_configuration.rb:178]: Can you add something to the pref name that makes it clear that the association is responsibility of the authentication system? Like `manually_associate_user_in_authentication_extension`, or even without the initial `manually_`.

Comment (elia) [reply core/lib/spree/app_configuration.rb:178]: Also the default (contrary to what I initially said 😅) should be `false`, to maintain the current behavior.

**Comment by spaghetticode** [comment]
@carlosep can you please review the commit message? It does not conform to the other ones on this project as it's very long and it does not read very nicely on github:

Screen Shot 2020-06-24 at 12 55 36 PM


I would keep the title at around 60 chars and add a description body with some details. Thanks 🙏

**Comment by carlosep** [comment]
@elia and @spaghetticode, all of the mentioned points were addressed, if something else is needed please reach out 
:slightly_smiling_face:

**Comment by elia** [comment]
@carlosep great job! I see the CI is failing because it treats deprecations as errors, I'm not sure how to deal with such a pervasive one, maybe @kennyadsl or @spaghetticode have handled similar ones in the past

**Comment by spaghetticode** [comment]
@carlosep you can remove the deprecations (and consequently their errors) by overriding the default configuration value for the new preference in [dummy_app.rb](https://github.com/solidusio/solidus/blob/master/core/lib/spree/testing_support/dummy_app.rb):

```ruby
Spree.config do |config|
  # ...
  config.use_combined_first_and_last_name_in_address = true
  config.use_legacy_order_state_machine = false
  # add your preference here:
  config.associate_user_in_authentication_extension = true
end
```

This way, when the test suite runs the preference value will be set to `true` instead of `false`.

**Comment by carlosep** [comment]
@spaghetticode, I added this configuration and still the CI broke. Is there some other place that requires this flag besides dummy_app.rb?

**Comment by kennyadsl** [comment]
Hey @carlosep, you can see another PR (like https://github.com/solidusio/solidus/pull/3749) as a reference for setting a "deprecation preference". Thanks!

**Comment by dmitriy-borowskiy** [comment]
Hi @elia, @carlosep @kennyadsl  Is still actual to check CI errors, fix and merge this PR?

**Comment by dborovsky** [comment]
Hi @elia, @carlosep @kennyadsl  Is still actual to check CI errors, fix and merge this PR?

**Comment by elia** [comment]
@dborovsky yes, this needs to be done, and IIRC this PR just lack the proper boilerplate for a smooth deprecation, interested in picking this up?

**Comment by dborovsky** [comment]
@elia  yea, I want to help with this  PR.

**Comment by elia** [comment]
That's great, if you have any question about the comments we left above just ask and I'll try to help 👍 
Also feel free to ping me on the solidus slack: http://slack.solidus.io/

**Comment by dborovsky** [comment]
@elia @spaghetticode  please review my PR https://github.com/solidusio/solidus/pull/4201

**Comment by waiting-for-dev** [comment]
Closing in favor of #4201

**Pull request closed without merging**